### DISCUSSION TITLE
How to handle 2-step AJAX authentication

### DISCUSSION THREAD

**Discussion by real-linarion** [django-otp/django-otp#145] [Q&A] [answered]
Hello,

I am currently a at my wits end and could use some guidance. My setup is as follows: I have a Django backend and a React frontend. Communication between the two is handled using DRF. For various reasons I want to use Django's Session backend, in the form of a cookie, for authentication. The user issues a login api call and Django issues the cookie in return. That much works fine. However, I also have to implement optional 2FA via email. I already use django-otp to do this on my admin site (thats the reason I went with this library way-back-when due to how easy it was to implement :) ). I would also very much like to use django-otp here.

So here is the problem as I see it. I cannot have the user supply their email token with their email and password, 2FA is supposed to stay optional. Therefore, I need two API endpoints, one for primary authentication with email and password and one for the email token. If the primary authentication is successfull and 2FA is enabled for that user, an email containing the token is sent. The frontend also gets a response, telling it, that the user will need to supply a token and navigates to another page where the user can input their token. If they do that and click submit, how will the backend know, that the user had previously completed the primary authentication step?

My first thought was: "Well it knows, because now the user has an email token stored in the database" And since that token is short lived and we are using rate limiting an attacker would likely not be able to guess the token during its life time and with the limited trys. But then I thought: "Well but what if I want to implement recovery codes in the future?" Well rate limiting would still be in place but technically the attacker could now have a very slow go at guessing the recovery codes, thus completely avoiding primary authentication. I'm also just feeling very iffy about this approach.

I think I'm majorly confused in regards to properly using django-otp so any help in clearing that up would be greatly appreciated.

**Comment by psagers** [comment]
If I'm understanding your question, it doesn't sound like there's anything especially novel introduced with the addition of 2FA.
Rate-limiting 2FA token verification is essentially the same problem as rate-limiting login attempts with a password: another case where an attacker has a theoretically unlimited amount of time to (slowly) test credentials. In practice, it should be less of an issue with 2FA tokens, as you can notify the user of successful logins, so if an attacker breaks the password, hopefully the user will reset it, at which point you should also revoke existing sessions.

1. Only verify tokens within authenticated sessions.
2. Rate-limit as appropriate.
3. Give the user tools to identify and respond to fraudulent logins.

**Reply by real-linarion** [reply]
Thank you for the reply :). The root of my problem lies in your first point:

> 1. Only verify tokens within authenticated sessions.

I only want to create an authenticated sessions AFTER the token has been verified. Only after the 2FA Challenge was completed, do I want to issue the user an authenticated session. I have a hunch that therein may lie my confusion 😅 i.e. that I'm conflating to separate things.

**Reply by psagers** [reply, marked as answer]
It sounds like you want to hold off calling [django.contrib.auth.login()](https://docs.djangoproject.com/en/5.0/topics/auth/default/#django.contrib.auth.login) until after 2FA is verified (for some users). In that case, keeping track of whether a given session is in the middle of the login flow is up to you. Something like `request.session['waiting-for-2fa'] = True` ought to do it.

Alternately, you could maintain the [authenticated/verified distinction](https://django-otp-official.readthedocs.io/en/stable/overview.html#authentication-and-verification) that django-otp provides. In that case, the initial login API would be available to anonymous users, the 2FA API would be restricted to authenticated users, and all other APIs would be limited to verified users. [django_otp.decorators.otp_required](https://django-otp-official.readthedocs.io/en/stable/auth.html#django_otp.decorators.otp_required) has an `if_configured` parameter so that it only applies to users with at least one device configured. This may not be exactly what you want, so the first approach should be fine if it makes more sense.

**Reply by real-linarion** [reply]
> It sounds like you want to hold off calling [django.contrib.auth.login()](https://docs.djangoproject.com/en/5.0/topics/auth/default/#django.contrib.auth.login) until after 2FA is verified (for some users).

Yes, exactly! That is what I want to achieve. Could you elaborate your first solution a bit? I thought that `request.session` would only come into play after `django.contrib.auth.login()` was called?

As to your second solution: I like that one alot, however, it would mean that I'd need to attach that decorator to all of my endpoints... I don't suppose that there is a way by which you could enforce the behavior of the `otp_required()` decorator with `if_configured=True` for all endpoints by default? Except for patching DRF's views I suppose...

Anyway thank you for clearing this up for me!

**Reply by psagers** [reply]
At this point, I think we’re just into standard Django stuff. The auth system uses `request.session` to keep track of the logged-in user, but you can use sessions to record any information or state you feel like. Just pick a probabilistically unique key and have at it. This is how views pass information to each other behind the scenes. In your case, you just need some key shared by the two auth APIs so the second one knows whether or not the session has already completed part 1.

Applying authorization policies to views is a broader topic, whether it’s decorators or class-based view hierarchies. I’m not familiar with DRF, but maybe it has hooks for custom authorization logic. Or you might write a custom authorization backend and implement the policy in terms of a permission (but beware: superusers will bypass this). Or maybe this is all too awkward and it’s best to stick with the two-phase authentication.

Side note: I’m using the terms “authentication” and “authorization” deliberately here. Authentication = verifying a user’s identity; authorization = deciding what a user is allowed to do. An important distinction that is sometimes elided.
